google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/Exceptions.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.bigquery.storage.v1;
+
+import com.google.api.gax.grpc.GrpcStatusCode;
+import com.google.common.collect.ImmutableMap;
+import com.google.protobuf.Any;
+import com.google.protobuf.InvalidProtocolBufferException;
+import javax.annotation.Nullable;
+
+/** Exceptions for Storage Client Libraries. */
+public final class Exceptions {
+  /** Main Storage Exception. Might contain map of streams to errors for that stream. */
+  public static class StorageException extends RuntimeException {
+
+    protected StorageException() {}
+
+    protected StorageException(String message) {
+      super(message);
+    }
+
+    protected StorageException(String message, Throwable cause) {
+      super(message, cause);
+    }
+
+    protected StorageException(Throwable cause) {
+      super(cause);
+    }
+
+    private ImmutableMap<String, GrpcStatusCode> errors;
+
+    public ImmutableMap<String, GrpcStatusCode> getErrors() {
+      return errors;
+    }
+
+    public void setErrors(ImmutableMap<String, GrpcStatusCode> value) {
+      this.errors = value;
+    }
+
+    private String streamName;
+
+    public String getStreamName() {
+      return streamName;
+    }
+
+    protected void setStreamName(String value) {
```

Comment:
I'd recommend removing these setters, and instead moving them to the constructor. This allows us to make this exception immutable which makes it safer and less foot-gun prone for our customers.

From a little poking I was able to come up with this minimized immutable base class which shows what I think we can do to trim things down, and make this a sealed and immutable. (I've also also added some more annotations for `@Nullable`)

```java
  public static class StorageException extends RuntimeException {

    private final String streamName;
    private final ImmutableMap<String, GrpcStatusCode> errors;

    private StorageException() {
      this(null, null, null, ImmutableMap.of());
    }

    private StorageException(
        @Nullable String message,
        @Nullable Throwable cause,
        @Nullable String streamName,
        ImmutableMap<String, GrpcStatusCode> errors) {
      super(message, cause);
      checkArgument(errors != null, "errors must be non null");
      this.streamName = streamName;
      this.errors = errors;
    }

    public ImmutableMap<String, GrpcStatusCode> getErrors() {
      return errors;
    }

    @Nullable
    public String getStreamName() {
      return streamName;
    }
  }
```

Reply:
Done. Changes in commit #8a2f95f